### PR TITLE
Use ragged input paged attention kernel for paged attention

### PR DESCRIPTION
**Pull request by liurupeng**
# Description

This change uses the ragged input paged attention kernel for both prefill&decode stages. I deliberately kept the kernel call for both stages so in future we could potentially use different kernel for different stages.

The rest of the description includes relevant details and context, examples:
- why is this change being made,
- the problem being solved and any relevant context,
- why this is a good solution,
- some information about the specific implementation,
- shortcomings of the solution and possible future improvements.

If the change fixes a bug or a Github issue, please include a link, e.g.,:
FIXES: b/123456
FIXES: #123456

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/inference/paged_attention_kernel_v2.py [hunk 1]
```diff
@@ -0,0 +1,721 @@
+# Copyright 2025 The JAX Authors.
```
Comment (xy12181) [inline MaxText/inference/paged_attention_kernel_v2.py:1]: should we import this file from https://github.com/jax-ml/jax/blob/main/jax/experimental/pallas/ops/tpu/ragged_paged_attention.py?

Comment (liurupeng) [reply MaxText/inference/paged_attention_kernel_v2.py:1]: The jax version that maxtext is using doesn't have the kernel yet, and migrate the jax version seems need some work

File: MaxText/inference/paged_attention.py [hunk 2]
```diff
@@ -125,25 +126,69 @@ def paged_dot_product_attention_with_max_and_sum(self, query, key, value):
...
+        kv_lens=page_state.sequence_lengths, #[max_num_seqs]
+        cu_q_lens=c_q_l, #[max_num_seqs+1]
+        page_indices=page_state.page_map, 
+        num_seqs=num_seqs,
+        num_kv_pages_per_block=self.pages_per_compute_block,
+    )
+    return jnp.expand_dims(result, axis=1)
 
-    d: kv_head_dim_size
+  def paged_attention_v1_decode(
```
Comment (patemotter) [inline MaxText/inference/paged_attention.py:183]: Is there any benefit to keeping the v1 code here?

File: MaxText/inference/paged_attention.py [hunk 3]
```diff
@@ -125,26 +126,67 @@ def paged_dot_product_attention_with_max_and_sum(self, query, key, value):
...
+    v_p = jnp.permute_dims(value_pages_var.value, (1, 2, 0, 3))
+    c_q_l = jnp.array([0, page_state.sequence_lengths[0]])
+    num_seqs = jnp.array([1])
+    query = query[0]  # [batch_size, max_num_tokens, num_kv_heads, head_dim] to [max_num_tokens, num_kv_heads, head_dim]
+    result = paged_attention_kernel_v2.ragged_paged_attention(
+        q=query,
+        k_pages=k_p,  # [total_num_pages, page_size, num_kv_heads, head_dim]
+        v_pages=v_p,
+        kv_lens=jnp.array([query.shape[0]]),  # max_prefill_length
+        cu_q_lens=c_q_l,
```
Comment (patemotter) [inline MaxText/inference/paged_attention.py:149]: What exactly are c_q_l and cu_q_lens? It's not very clear just from the variable naming.

File: MaxText/inference/paged_attention.py [hunk 4]
```diff
@@ -46,6 +47,7 @@
 D_KV = common_types.D_KV
 
 shard_map = shard_map.shard_map
+use_kernel_v2 = False
```
Comment (patemotter) [inline MaxText/inference/paged_attention.py:50]: 1. Do we want to default to this being False? 
2. If this variable is going to be in use for a while then we should add it to the actual config in base.yml.

Comment (liurupeng) [reply MaxText/inference/paged_attention.py:50]: we could set it to true, my worry is that the new kernel doesn't support SPMD, folks who use > 1 host TPU slice will have issue using this. For 2, the long term goal might be having a flag to choose which kernels to use, but this flag is only temporary.

File: MaxText/inference/paged_attention.py (hunk 3 above)
Comment (liurupeng) [reply MaxText/inference/paged_attention.py:149]: SG! will update this

File: MaxText/inference/paged_attention.py (hunk 2 above)
Comment (liurupeng) [reply MaxText/inference/paged_attention.py:183]: From XLA team's benchmarking, this kernel has 20% performance gain than v2 kernel in "decode only" tasks. So we could keep this in case we use different kernels for different stages

Comment (patemotter) [reply MaxText/inference/paged_attention.py:183]: Sounds good, maybe add that as a comment that v2 may not always be the best choice.

File: MaxText/inference/paged_attention.py (hunk 4 above)
Comment (patemotter) [reply MaxText/inference/paged_attention.py:50]: That's fine, let's default to it being false until we feel more comfortable with the state of things.

File: MaxText/inference/paged_attention.py (hunk 2 above)
Comment (liurupeng) [reply MaxText/inference/paged_attention.py:183]: will do

File: MaxText/inference/paged_attention.py [hunk 5]
```diff
@@ -125,26 +127,68 @@ def paged_dot_product_attention_with_max_and_sum(self, query, key, value):
...
-    Annotations:
-    b: batch_size
-    s: sequence length
-    n: query heads
+    """Apply ragged input Paged Attention in prefill only."""
+    k_p = jnp.permute_dims(key_pages_var.value, (1, 2, 0, 3))
+    v_p = jnp.permute_dims(value_pages_var.value, (1, 2, 0, 3))
+    c_q_l = jnp.array([0, page_state.sequence_lengths[0]])  # [0, prefill_true_length]
+    num_seqs = jnp.array([1])
+    query = query[0]  # [batch_size, max_num_tokens, num_kv_heads, head_dim] to [max_num_tokens, num_kv_heads, head_dim]
```
Comment (wyzhang) [inline MaxText/inference/paged_attention.py:143]: nit: assert batch_size == 1?

File: MaxText/inference/paged_attention.py [hunk 6]
```diff
@@ -125,26 +127,68 @@ def paged_dot_product_attention_with_max_and_sum(self, query, key, value):
...
       page_state: page_manager.PageState,
   ) -> Array:
-    """Apply Paged Attention.
-
-    Annotations:
-    b: batch_size
-    s: sequence length
-    n: query heads
+    """Apply ragged input Paged Attention in prefill only."""
+    k_p = jnp.permute_dims(key_pages_var.value, (1, 2, 0, 3))
```
Comment (wyzhang) [inline MaxText/inference/paged_attention.py:139]: nit: Mind adding a comment for the expected shape of key_pages_var

num_kv_heads, num_pages, tokens_per_page, kv_head_dim_size

File: MaxText/inference/paged_attention.py [hunk 7]
```diff
@@ -125,26 +127,68 @@ def paged_dot_product_attention_with_max_and_sum(self, query, key, value):
...
+    query = query[0]  # [batch_size, max_num_tokens, num_kv_heads, head_dim] to [max_num_tokens, num_kv_heads, head_dim]
+    result = paged_attention_kernel_v2.ragged_paged_attention(
+        q=query,
+        k_pages=k_p,  # [total_num_pages, page_size, num_kv_heads, head_dim]
+        v_pages=v_p,
+        kv_lens=jnp.array([query.shape[0]]),  # max_prefill_length
+        cu_q_lens=c_q_l,  # the accumulative real lengths of requests, starting from 0
+        page_indices=page_state.page_map,
+        num_seqs=num_seqs,
+        # num_kv_pages_per_block=self.pages_per_compute_block, #I saw repeated response when enabled this
```
Comment (wyzhang) [inline MaxText/inference/paged_attention.py:152]: nit: Add a TODO. Move the comment above the line of the code.

File: MaxText/inference/paged_attention.py [hunk 8]
```diff
@@ -125,26 +127,68 @@ def paged_dot_product_attention_with_max_and_sum(self, query, key, value):
...
+    )
+    return jnp.expand_dims(result, axis=0)
```
Comment (wyzhang) [inline MaxText/inference/paged_attention.py:154]: nit: Mind adding a comment for the shape of the result before expand_dims

**Review by wyzhang — APPROVED** [review]

File: MaxText/inference/paged_attention.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (liurupeng) [reply MaxText/inference/paged_attention.py:50]: thanks! updated

File: MaxText/inference/paged_attention.py (hunk 6 above)
Comment (liurupeng) [reply MaxText/inference/paged_attention.py:139]: updated!

File: MaxText/inference/paged_attention.py (hunk 5 above)
Comment (liurupeng) [reply MaxText/inference/paged_attention.py:143]: udpated!

File: MaxText/inference/paged_attention.py (hunk 7 above)
Comment (liurupeng) [reply MaxText/inference/paged_attention.py:152]: Sg, updated!

File: MaxText/inference/paged_attention.py (hunk 8 above)
Comment (liurupeng) [reply MaxText/inference/paged_attention.py:154]: updated, thanks!

**Review by gobbleturk — APPROVED** [review]

**Review by vipannalla — APPROVED** [review]
